nerfstudio/configs/annotated_types.py

```diff
@@ -0,0 +1,139 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dataclasses
+from collections import OrderedDict
+from typing import TYPE_CHECKING
+
+import tyro
+
+from nerfstudio.configs.dataparser_configs import dataparser_configs
+from nerfstudio.configs.external_methods import get_external_methods
+from nerfstudio.configs.method_configs import descriptions as method_descriptions
+from nerfstudio.configs.method_configs import method_configs
+from nerfstudio.data.dataparsers.base_dataparser import DataParserConfig
+from nerfstudio.engine.trainer import TrainerConfig
+from nerfstudio.plugins.registry import discover_dataparsers, discover_methods
+
+
+def merge_methods(methods, method_descriptions, new_methods, new_descriptions, overwrite=True):
+    """Merge new methods and descriptions into existing methods and descriptions.
+    Args:
+        methods: Existing methods.
+        method_descriptions: Existing descriptions.
+        new_methods: New methods to merge in.
+        new_descriptions: New descriptions to merge in.
+    Returns:
+        Merged methods and descriptions.
+    """
+    methods = OrderedDict(**methods)
+    method_descriptions = OrderedDict(**method_descriptions)
+    for k, v in new_methods.items():
+        if overwrite or k not in methods:
+            methods[k] = v
+            method_descriptions[k] = new_descriptions.get(k, "")
+    return methods, method_descriptions
+
+
+def sort_methods(methods, method_descriptions):
+    """Sort methods and descriptions by method name."""
+    methods = OrderedDict(sorted(methods.items(), key=lambda x: x[0]))
+    method_descriptions = OrderedDict(sorted(method_descriptions.items(), key=lambda x: x[0]))
+    return methods, method_descriptions
+
+
+# DataParsers
+external_dataparsers, _external_dataparsers_help = discover_dataparsers()
+all_dataparsers = {**dataparser_configs, **external_dataparsers}
+
+if TYPE_CHECKING:
+    # For static analysis (tab completion, type checking, etc), just use the base
+    # dataparser config.
+    DataParserUnion = DataParserConfig
+else:
+    # At runtime, populate a Union type dynamically. This is used by `tyro` to generate
+    # subcommands in the CLI.
+    DataParserUnion = tyro.extras.subcommand_type_from_defaults(
+        all_dataparsers,
+        prefix_names=False,  # Omit prefixes in subcommands themselves.
+    )
+
+AnnotatedDataParserUnion = tyro.conf.OmitSubcommandPrefixes[DataParserUnion]  # Omit prefixes of flags in subcommands.
+"""Union over possible dataparser types, annotated with metadata for tyro. This is
+the same as the vanilla union, but results in shorter subcommand names."""
+
+
+def fix_method_dataparser_type(method_config: TrainerConfig):
```

Comment:
@brentyi , can you please take a closer look at this function?

Reply:
This makes sense but feels scary to me, is there no way to keep the parsers annotated as `AnnotatedDataParserUnion`? Is the issue with circular imports? Would breaking this file up into smaller files (like one where the dataparser union is defined and one where the method union is defined) help?

Reply:
I think having the AnnotatedDataParserUnion in the DataManagerConfig is a bad design which leads to circular imports when you want to define the DataParser and something else (trainerconfig) in the same file. I agree that redefining the types here is not the safest. I think it would be preferable if there was a way to instruct tyro to do it instead: replace base types with the annotated unions? What do you think? If you like this solution I can try to implement it in tyro.

Reply:
Sorry for missing your reply, I had accidentally left my comment twice and assumed you hadn't replied yet because I was looking at the other one...

> I think it would be preferable if there was a way to instruct tyro to do it instead: replace base types with the annotated unions

Yeah, this makes sense. There's an unfinished PR that would enable this here: https://github.com/brentyi/tyro/pull/30

If you're willing to make a PR to tyro making this kind of thing possible — either building on my PR or starting from scratch — that'd of course be awesome, but I don't have a good sense of how long something like that would take (or how long getting ramped up on the tyro codebase would take).

If it's too much effort just splitting things like the DataParser + trainerconfig that you described into separate files seems like an unideal but workable short-term solution.